### PR TITLE
Range for freeActions

### PR DESCRIPTION
**Pull request by rzulian**
This will allow the viewer to allow multiple payments

### CODE REVIEW

**Comment by coyotte508** [comment]
Maybe instead of adding multiples of free action, you can add a range parameter (in addition to income & cost) which tells the UX how many times the action can be exectured. then let the UX handle it. Maybe we can make the buttons dropdowns with how many times (and the burn button too).

**Comment by coyotte508** [comment]
Here are dropdowns in bootstrap: https://getbootstrap.com/docs/4.0/components/dropdowns/

**Comment by rzulian** [comment]
Nice idea! This PR is adding the range. Please have a look and approve.I'll work on the viewer then

File: src/player.ts [hunk 1]
```diff
@@ -109,6 +109,21 @@ export default class Player extends EventEmitter {
     return true;
   }
 
+  maxPayRange(reward: Reward[]): number {
+    const rewards = Reward.merge(reward);
+
+    let max = 0;
+    while (max >= 0) {
```
Comment (coyotte508) [inline src/player.ts]: Can be replaced by `while (1)`

File: src/player.ts [hunk 2]
```diff
@@ -109,6 +109,21 @@ export default class Player extends EventEmitter {
...
+      for (const rew of rewards) {
+        if (!this.data.hasResource(new Reward( rew.count * (max + 1), rew.type ))) {
+          return max;
+        }
+      }
+      max += 1;
+    }
+    return max;
```
Comment (coyotte508) [inline src/player.ts]: unreachable code, can be removed

**Comment by coyotte508** [comment]
By the way, bootstrap, in addition to regular dropdowns, has a "split dropdown", maybe we can use that on the UX. With the main part of the button doing the regular freeaction, and the dropdown part allowing the choice to do the free action 1-X times.

File: src/player.ts (hunk 1 above)
Comment (coyotte508) [reply src/player.ts]: actually it can be replaced by `for (let max = 0; ; max += 1)`

**Pull request merged**